The [opa eval] command is unable to access Rego policies located on other drives on windows

pckvcode opened the pull request:
The [opa eval] command is unable to access Rego policies located on other drives in Windows. For more details, see https://github.com/open-policy-agent/opa/issues/6910 and https://github.com/open-policy-agent/conftest/issues/979.

OPA should be capable of accessing Rego policies from specified paths including drives (e.g., c:\a\b\c.rego).

The code has been updated to allow OPA to load Rego policies from paths with drives. Now, OPA can load Rego policies from paths(e.g., `/a/b/c/example.rego`), path with URLs(e.g., `file:///path/to/file.json`), and path with drives(e.g., `C:\a\b\c\example.rego`).

Fixes https://github.com/open-policy-agent/opa/issues/6910

<!--

Thanks for submitting a PR to OPA!

Before pressing 'Create pull request' please read the checklist below.

* All code changes should be accompanied with tests. If you are not
modifying any tests, just provide a short explanation of why updates
to tests are not necessary. In addition to helping catch bugs, tests
are extremely helpful in providing _context_ that explains how your
changes can be used.

* All changes to public APIs **must** be accompanied with
docs. Examples of public APIs include built-in functions,
config fields, and of course, exported Go types/functions/constants/etc.

* Commit messages should explain _why_ you made the changes, not what
you changed. Use active voice. Keep the subject line under 50
characters or so.

* All commits must be signed off by the author. If you are not
familiar with signing off, see our contributor guide below.

For more information on contributing to OPA see:

* [Contributing Guide](https://www.openpolicyagent.org/docs/latest/contributing/)
  for high-level contributing guidelines and development setup.

-->

### Why the changes in this PR are needed?
The `opa eval` command encountered an issue where it could not load Rego policies from different drives on Windows. The error message was:

```
CreateFile \\Users\\pck\\AppData\\Local\\Temp\\temp_dir_12342318055089\\example.rego: The system cannot find the path specified
```
It fixes: https://github.com/open-policy-agent/opa/issues/6910 and https://github.com/open-policy-agent/conftest/issues/979


<!--
Include a short description of WHY the changes were made.
-->

### What are the changes in this PR?

#### Code Update:
The code has been modified to allow OPA to load Rego policies from paths that include drive letters. Now, OPA supports loading Rego policies from:
- Paths (e.g., `/a/b/c/example.rego`)
- Paths with URLs (e.g., `file:///path/to/file.json`)
- Paths with drive letters (e.g., `C:\a\b\c\example.rego`)

#### Documentation Reference:
According to the [official documentation](https://www.openpolicyagent.org/docs/latest/cli/#examples), the expected values for the `--data` flag in the `opa eval` command are:
1. A simple path (e.g., `/a/b/c/example.rego`)
2. A path with a URL (e.g., `file:///path/to/file.json`)

Ideally, specifying a path like `C:\a\b\c\example.rego` should work, but it was failing. The code was modified to support paths with drive letters by ensuring that the prefix is only separated when its length is greater than 1.

<!--
Include a short description of WHAT changes were made.
-->

### Notes to assist PR review:

#### Steps to Reproduce the issue:
1. Place the OPA binary on the D drive.
2. Store the Rego policies on the C drive.
3. Attempt to evaluate the policies using the `opa eval` command with the policies located on the C drive. The above error will occur. For more details, refer to [this GitHub issue](https://github.com/open-policy-agent/opa/issues/6910#issuecomment-2282810960).


<!--
Here you can add information you think will help the reviewer(s).
-->

### Further comments:

#### [opa eval] command before fix. (Failed: opa in D drive and Failed to load rego policies from C drive.)
![Screenshot 2024-08-11 at 9 10 42 PM_Edited](https://github.com/user-attachments/assets/9ef23efc-032b-4a96-86d1-5885f67b3cac)


#### [opa eval] command after fix. (Success: opa in D drive and Failed to load rego policies from C drive.)
<img width="1164" alt="Screenshot 2024-08-12 at 1 50 03 PM" src="https://github.com/user-attachments/assets/9b07b095-170a-453e-b01f-9bf46ba39724">

#### Unit Testcase pass screenshot. [Function tested](https://github.com/open-policy-agent/opa/blob/main/loader/loader.go#L568)
<img width="644" alt="Screenshot 2024-08-12 at 1 50 28 PM" src="https://github.com/user-attachments/assets/4ae0a2fc-142e-4b0b-a900-b5af74e32656">

<!--
Here you can include links to additional resources related to the changes, discuss your solution, other approaches you considered etc.
-->

 @boranx Can you please review this change and share your comments? Thanks in advance

philipaconrad commented:
The new test appears to be [causing failures](https://github.com/open-policy-agent/opa/actions/runs/10348715262/job/28641988120?pr=6922#step:6:252) on non-Windows platforms (`linux`, `darwin`). We may need to either split the test off for Windows, or find some other way to OS-gate it.

pckvcode commented:
@philipaconrad 
I fixed it now. Can you please approve the workflow and review the changes?

I got to know we are supporting these inputs.
- Paths (e.g., /a/b/c/example.rego)
- Paths with URLs (e.g., file:///path/to/file.json)
- Paths with drive letters and URLs (e.g., "C:file:///C:/a/b/c")

Added the support for below

- Paths with drive letters (e.g., C:\a\b\c\example.rego)

My comments:
- C:file:///C:/a/b/c is not a right url format but we are supporting it. no where mentioned in the documentation https://www.openpolicyagent.org/docs/latest/cli/#examples
- Made code changes to support all the path types mentioned above.

boranx reviewed (commented):
IMO, intuitively, I'd say we can support the following
```
Paths (e.g., /a/b/c/example.rego)
Paths with URLs (e.g., file:///path/to/file.json)
```
but I'm not sure if `Paths with drive letters and URLs (e.g., "C:file:///C:/a/b/c")` is a valid format, as I haven't seen such thing before 🤔 

My 2 cents is to add a unit test for each one of those cases, as supported vs not supported

pckvcode commented:
I tested with the latest OPA Windows AMD64 binary. Here are the results:

```bash
curl -O https://github.com/open-policy-agent/opa/releases/download/v0.67.1/opa_windows_amd64.exe
```
**Note:** 
- In all these cases, I kept the binary on the Windows D drive.

**Succeeded for these paths:**
- Paths with URLs (e.g., `file:///path/to/file.json`)
- Paths with drive letters and URLs (e.g., `"C:file:///C:/a/b/c"`)
- Standard paths (e.g., `/a/b/c/example.rego`)

```bash
Commands used:

.\opa_windows_amd64.exe eval --input "C:\Users\pck\Punith\TestFiles\input.json" --data "file:///C:/Users/pck/Punith/TestFiles/rego_policies/example.rego" "data.example.authz.allow"
.\opa_windows_amd64.exe eval --input "C:\Users\pck\Punith\TestFiles\input.json" --data "C:file:///C:/Users/pck/Punith/TestFiles/rego_policies/example.rego" "data.example.authz.allow"
# able to access from D drive since the binary in D drive
.\opa_windows_amd64.exe eval --input "C:\Users\pck\Punith\TestFiles\input.json" --data "D:/punith/opa-bug/TestFiles/rego_policies/example.rego" "data.example.authz.allow"
```
<img width="1108" alt="Screenshot 2024-08-23 at 1 52 16 PM" src="https://github.com/user-attachments/assets/3235b320-66dd-4343-812b-3da20c8e7998">
<img width="1101" alt="Screenshot 2024-08-23 at 1 52 45 PM" src="https://github.com/user-attachments/assets/4373f9b5-d31c-4def-8ee5-16de9736d953">
<img width="1099" alt="Screenshot 2024-08-23 at 1 54 15 PM" src="https://github.com/user-attachments/assets/60c35794-a3a0-4bcf-9dd5-135c4484a36a">


**Failed for this path:**
- Paths with drive letters (e.g., `C:\a\b\c\example.rego`)

```bash
# Fails since the binary is in D drive and loading policies from C drive
.\opa_windows_amd64.exe eval --input "C:\Users\pck\Punith\TestFiles\input.json" --data "C:\Users\pck\Punith\TestFiles\rego_policies\example.rego" "data.example.authz.allow"
```
<img width="1106" alt="Screenshot 2024-08-23 at 1 54 48 PM" src="https://github.com/user-attachments/assets/a5260541-16cb-4ea2-84ed-5a3ceb1d20b9">

----------

I then tested with my code changes, and it worked for the following paths:

- Paths with URLs (e.g., `file:///path/to/file.json`)
- Paths with drive letters and URLs (e.g., `"C:file:///C:/a/b/c"`)
- Standard paths (e.g., `/a/b/c/example.rego`)
- Paths with drive letters (e.g., `C:\a\b\c\example.rego`)

Commands used:

```bash
.\opa-new.exe eval --input "C:\Users\pck\Punith\TestFiles\input.json" --data "file:///C:/Users/pck/Punith/TestFiles/rego_policies/example.rego" "data.example.authz.allow"
.\opa-new.exe eval --input "C:\Users\pck\Punith\TestFiles\input.json" --data "C:file:///C:/Users/pck/Punith/TestFiles/rego_policies/example.rego" "data.example.authz.allow"
.\opa-new.exe eval --input "D:/punith/opa-bug/TestFiles/rego_policies/example.rego" "data.example.authz.allow"
.\opa-new.exe eval --input "C:\Users\pck\Punith\TestFiles\input.json" --data "C:\Users\pck\Punith\TestFiles\rego_policies\example.rego" "data.example.authz.allow"
```

<img width="1115" alt="Screenshot 2024-08-23 at 1 55 12 PM" src="https://github.com/user-attachments/assets/d1f64296-a004-405e-ac5c-b7fddcb83aaa">
<img width="1096" alt="Screenshot 2024-08-23 at 1 55 38 PM" src="https://github.com/user-attachments/assets/4d0cafe2-1919-45b7-9c30-e8d9d95cb250">
<img width="1092" alt="Screenshot 2024-08-23 at 1 56 02 PM" src="https://github.com/user-attachments/assets/753e9f7d-27be-499f-ab6a-ed3445cc02c8">
<img width="1097" alt="Screenshot 2024-08-23 at 1 56 39 PM" src="https://github.com/user-attachments/assets/f9d187a6-1d5e-44e3-b8f7-47d827b9a2b6">

---------
I also added a unit test case to verify loading files from paths with drive letters (e.g., `C:\a\b\c\example.rego`).

johanfylling commented on loader/loader.go:
```diff
@@ -582,7 +582,9 @@ func SplitPrefix(path string) ([]string, string) {
 	parts := strings.SplitN(path, ":", 2)
 	if len(parts) == 2 && len(parts[0]) > 0 {
-		return strings.Split(parts[0], "."), parts[1]
+		if strings.HasPrefix(parts[1], "file:///") || len(parts[0]) > 1 {
```
Why do we put significance on the path containing `file:///` or the document prefix being more than one chars long for us to return those parts rather than the given `path`?

johanfylling replied:
To add some context: I believe the prefix part is the prefix for which the contents of that file will be found under the `data` document during evaluation.

E.g.: given the following data file `data.json`:

```json
{
  "foo": 42
}
```

and we run: 

```cmd
$ opa eval -fpretty -d file:///path/to/my/file/data.json 'data = x'
```

we expect the output: 
```cmd
+------------+
|     x      |
+------------+
| {"foo":42} |
+------------+
```

But if we add the `a:` prefix to the file path, we expect this prefix to be added to the data document:

```cmd
opa eval -fpretty -d a:file:///path/to/my/file/data.json 'data = x'
```

```cmd
+------------------+
|        x         |
+------------------+
| {"a":{"foo":42}} |
+------------------+
```

johanfylling commented on loader/loader_test.go:
```diff
@@ -879,6 +889,15 @@ func TestSplitPrefix(t *testing.T) {
 		},
+		{
+			input:    "c:/a/b/c",
+			wantPath: "c:/a/b/c",
```
This doesn't seem compatible with how data prefixes work in OPA, as we'd expect `c` to become the data prefix.

johanfylling reviewed (commented):
Just a few questions.

I don't have a windows machine to test on, but I would expect the path `file:///c:/path/to/my/file/data.json` to work. Have you tested this format?

pckvcode commented:
> Just a few questions.
> 
> I don't have a windows machine to test on, but I would expect the path `file:///c:/path/to/my/file/data.json` to work. Have you tested this format?

I tried using the path `file:///c:/path/to/my/file/data.json`, and it worked. 
However, when I used `c:/path/to/my/file/data.json`, it failed. 
<img width="1109" alt="Screenshot 2024-08-27 at 6 21 16 PM" src="https://github.com/user-attachments/assets/404e6648-7bac-4239-aee3-e85a330c898c">
-------
<img width="1051" alt="Screenshot 2024-08-27 at 6 30 39 PM" src="https://github.com/user-attachments/assets/926a2d22-5d8b-4191-ae3b-1883a757681d">

--------

Is it advisable to always use a file URL, especially on Windows? 
How does the system behave if we provide the path as `c:/path/to/my/file/data.json?` In this case, should c be interpreted as a prefix or as the drive letter?

johanfylling commented:
The current behavior is to interpret the `c:` in `c:/path/to/my/file/data.json` as the data prefix `c`. Modifying this behavior would be a breaking change, and given that there are no recommendations/requirements for the number of characters in a prefix, I don't think this is something we can safely change.

Given the verbosity of file-urls and that Windows is the only supported operating system that uses the drive letter syntax, I don't think we can say the usage of file-urls are a general recommendation; but definitely a requirement for specifying absolute paths on Windows. 
The `prefix:path` syntax is an unfortunate overlap. If not already clarified in the docs, I think it warrants an mention. 

Thank you for bringing this to our attention @pckvcode 😃! I'm closing this PR, but if you still want to contribute to the project with your findings, you are more than welcome to open a new PR with updates to the docs.
